File: Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        hoge = []
        fuga = {")": "(", "]": "[", "}": "{"}
        for i in s:
            if i in "([{": hoge.append(i)
            else:
                if len(hoge) and fuga[i] == hoge[-1]: hoge.pop()
                else: return False
        return len(hoge) == 0

File: C++/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for (char i: s) {
            if (i == '(' || i == '[' || i == '{') stk.push(i);
            else {
                if (stk.empty()) return false;
                if (i == ')' && stk.top() != '(' || i == ']' && stk.top() != '[' || i == '}' && stk.top() != '{')
                    return false;
                stk.pop();
            }
        }

        return stk.empty();
    }
};
